builder/vsphere/common/step_boot_command.go

```diff
@@ -68,17 +70,60 @@ func (s *StepBootCommand) Run(ctx context.Context, state multistep.StateBag) mul
 		pauseFn = state.Get("pauseFn").(multistep.DebugPauseFn)
 	}
 
-	port := state.Get("http_port").(int)
+	var ip string
+	var err error
+	port, ok := state.Get("http_port").(int)
+	if !ok {
+		ui.Error("error retrieving 'http_port' from state")
+		return multistep.ActionHalt
+	}
+
 	if port > 0 {
-		ip := state.Get("http_ip").(string)
-		s.Ctx.Data = &bootCommandTemplateData{
-			ip,
-			port,
-			s.VMName,
+		keys := []string{"http_bind_address", "http_interface", "http_ip"}
+
+		for _, key := range keys {
+			value, ok := state.Get(key).(string)
+			if ok && value != "" {
+				switch key {
+				case "http_bind_address":
+					ip = value
+					ui.Sayf("Using IP address %s from 'http_bind_address'.", ip)
+				case "http_interface":
+					ip, err = hostIP(value)
+					if err != nil {
+						err = fmt.Errorf("error determining IP address from interface %s: %s", value, err)
+						state.Put("error", err)
+						ui.Error(err.Error())
+						return multistep.ActionHalt
+					}
+					ui.Sayf("Using IP address %s from 'http_interface' %s.", ip, value)
+				case "http_ip":
+					ip = value
+					httpIPSetByUser, ok := state.Get("http_ip_set_by_user").(bool)
+					if ok && httpIPSetByUser {
+						ui.Sayf("Using IP address %s from 'http_ip'.", ip)
+					} else {
+						ui.Sayf("Using IP address %s from first non-loopback interface.", ip)
+					}
```

Comment:
Is it really necessary to read "http_ip_set_by_user" from state and display a custom message if using the default loopback. I was a little confused by this extra check because I would assume if `state.Get("http_ip_set_by_user").(bool)` is true then we know that `http_ip` is being used so this case can be a guard clause above the switch statement. 

Below is a suggested rewrite for the for loop. 
```go
	keys := []string{"http_bind_address", "http_interface", "http_ip"}
	for _, key := range keys {
		value, ok := state.Get(key).(string)
		if ok && value == "" {
			continue
		}

		switch key {
		case "http_bind_address":
			ip = value
			ui.Sayf("Using IP address %s from 'http_bind_address'.", ip)
		case "http_interface":
			ip, err = hostIP(value)
			if err != nil {
				err = fmt.Errorf("error determining IP address from interface %s: %s", value, err)
				state.Put("error", err)
				ui.Error(err.Error())
				return multistep.ActionHalt
			}
			ui.Sayf("Using IP address %s from 'http_interface' %s.", ip, value)
		case "http_ip":
			ip = value
			ui.Sayf("Using IP address %s from 'http_ip'.", ip)
		}
		ui.Sayf("Using IP address %s from %q.", ip, key)
		break
	}
```

Reply:
Resolved in 53f1531.